docs/Extended_dry_run.rst

```diff
@@ -0,0 +1,731 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To make runtime patching of files in easyblocks easier, and to do it with taking the possibility of being in dry run
+module into account, the EasyBuild framework provides the ``apply_regex_substitutions`` function (available from the
+``easybuild.tools.filetools`` module, since EasyBuild v2.4.0).
+
+This function takes two arguments: a path to the file that should be patched, and a list of tuples specifying the regular
+expression pattern to match on, and the string value that should be used as replacement text.
+
+For example (simple fictional example)::
+
+    # replace value for C++ compiler
+    apply_regex_substitutions('config.mk', [('^(CPLUSPLUS\s*=).*', '\1 %s' % os.environ['CXX'])])
+
+When used in dry run mode, it will produce a message like::
+
+    applying regex substitutions to file config.mk
+      * regex pattern '^(CPLUSPLUS\s*=\s).*', replacement string '\1 g++'
+
+.. _extended_dry_run_guidelines_easyblocks_framework_functions_run_cmd:
+
+Executing commands: ``run_cmd`` and ``run_cmd_qa``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To execute shell commands, the ``run_cmd`` and ``run_cmd_qa`` functions are provided by the EasyBuild framework in the
+``easybuild.tools.run`` module, with the latter providing support for running interactive commands.
+
+In their simplest form, they simply take the command to execute as a string. For example::
+
+      run_cmd("tcsh ./compile -j %s wrf" % self.cfg['parallel'])
+
+In dry run mode, these functions just produce a dry run message, rather than actually executing the specified command.
+For example::
+
+      running command "tcsh ./compile -j 4 wrf"
+      (in /home/example/eb/software/WRF/3.6.1-intel-2015a-dmpar/WRF-3.6.1)
+
+Take into account that the directory included in the message may not be 100% accurate,
+see :ref:`extended_dry_run_overview_build_dir_guess`.
+
+Silently executing commands
+###########################
+
+The ``verbose`` named argument supported by the ``run_cmd`` function allows to execute a particular command silently,
+i.e. without producing a dry run message.
+
+For example::
+
+    # only run for debugging purposes
+    run_cmd("ulimit -v", verbose=False)
+
+Forced execution of particular commands
+#######################################
+
+Sometimes, it can be required that specific (light-weight) commands are *always* executed, because they have
+side-effects that are assumed to have taken place later in the easyblock.
+
+For this, the ``run_cmd`` function support another named argument, i.e. ``forced``.
+When set to ``True``, the specified command will always be executed, even when in dry run mode.
+
+This is mainly intended for use in the EasyBuild framework itself, where commands that verify certain things must
+be executed, but it can also be useful for easyblocks (if used correctly).
+
+For example::
+
+    out, exit_code = run_cmd("type module", simple=False, forced=True)
```

Comment:
not necessarily a documentation issue, but I don't like the parameter being called "forced." Who knows if there will be a future even more forced option.

Maybe it should be "run_in_dry" or "forced_wet" :)